### ISSUE TITLE
Support of local environmental variable on Ansible Controller that populated from ansible variables for kubectl connection plugin 

### ISSUE THREAD

**Issue by yurnov** [ansible-collections/kubernetes.core#698]
##### SUMMARY
`kubernetes.core.kubectl` connection plugin can be used to execute the content of plays inside pods in the Kubernetes cluster, it's easy to use an efficient tool. But sometimes clusters require local environmental variables for authorization, i.e. `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` for authorization, even we have a `kubeconfig`.

So, it's will be nice to have support of local environment variable for kubectl connection plug-in. Importantly, I talking about the local variable that required for `kubectl` binary (technically `aws` binary in case of AWS, but other cases may require another enviromental variable for another cluster type), not for remote enviromental variable that is natively supported by Ansible.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
kubernetes.core.kubectl connection plugin

##### ADDITIONAL INFORMATION

In some cases `kubectl` require not only `kubeconfig` file, but some additional credentionals (i.e. platform/cloud specific) for anothentication/authorization, and enviromental variable is conviniant place to store them, i.e. `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` in case of AWS EKS with kubeconfig that invoke aws binary. AWS is only one example, the similar can be conviniant for another cloud providers as well. 

It's not a issue in case if Ansible is used to manage something just in one cluster, but in case of multiple clusters mentioned in inventory, `kubectl` connection plug can't be used.

So, this issue it for request support of local environment variable, to have working something like:
```
- hosts: cluster1:cluster2
  gather_facts: no
  any_errors_fatal: yes
  vars:
    ansible_connection: "kubernetes.core.kubectl"
    ansible_kubectl_namespace: "test"
    ansible_kubectl_config: "{{ kubeconfig }}"
    ansible_kubectl_pod: "shell"
    ansible_kubectl_container: "shell"
    ansible_kubectl_local_env:
       AWS_ACCESS_KEY_ID: "{{ aws_access_key_id }}"
       AWS_SECRET_ACCESS_KEY: "{{ aws_secret_access_key }}"
  environment:
    TEST_ENV1: value1
    TEST_ENV2: value2

  tasks:
  - name: test
    ansible.builtin.shell: env
    register: result
  - debug:
      var: result.stdout_lines
```

Expected result for this snippet is getting enviromental variable on running container in two clusters.
Actual result (as on version 3.0.1 of kubernetes.core collection) is failing connection at least to one cluster.

It's important specifically for kubernetes.core.kubectl to have ability to use same playbook in multi-cluster (and multi public cluster provider) scenario.

Update:

to be fully clear, this issue is request of implementation support of:
```
    ansible_kubectl_local_env:
       AWS_ACCESS_KEY_ID: "{{ aws_access_key_id }}"
       AWS_SECRET_ACCESS_KEY: "{{ aws_secret_access_key }}"
```
from a example above, and it's a __local__ environmental variable on the ansible controller, not on on target host

**Comment by yurnov** [comment]
I tried to test [this way](https://github.com/ansible-collections/kubernetes.core/compare/main...yurnov:kubernetes.core:attempt_to_add_local_shell_args#diff-364de5605d15a075ec88200261e48ea628916f64506c8159a2902049bea1d22bR308) with test playbook, but it doesn't work:
```
root@ubuntu-shell:/# cat test2.yaml
- hosts: localhost
  gather_facts: no
  any_errors_fatal: yes
  vars:
    ansible_connection: "kubectl"
    ansible_kubectl_namespace: "test"
    ansible_kubectl_config: "/.kube/config"
    ansible_kubectl_pod: "ubuntu"
    ansible_kubectl_container: "ubuntu"
    ansible_kubectl_local_shell_args: "EKS_TOKEN=test"
  environment:
    TEST_ENV1: value1
    TEST_ENV2: value2

  tasks:
  - name: test
    ansible.builtin.shell: env
    register: result
  - debug:
      var: result.stdout_lines
root@ubuntu-shell:/# ansible-playbook -vvv test2.yaml
ansible-playbook [core 2.14.5]
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.10/dist-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/local/bin/ansible-playbook
  python version = 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (/usr/bin/python3)
  jinja version = 3.1.3
  libyaml = True
No config file found; using defaults
host_list declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
script declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
auto declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
yaml declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
ini declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
toml declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'
Skipping callback 'default', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.

PLAYBOOK: test2.yaml ***************************************************************************************************************************
1 plays in test2.yaml

PLAY [localhost] *******************************************************************************************************************************

TASK [test] ************************************************************************************************************************************
task path: /test2.yaml:16
redirecting (type: connection) ansible.builtin.kubectl to kubernetes.core.kubectl
<127.0.0.1> ESTABLISH kubectl CONNECTION
<127.0.0.1> EXEC ['/usr/bin/bash', '-c', 'EKS_TOKEN=test', '', '&&', '', '/usr/local/bin/kubectl', '-n', 'test', '--kubeconfig', '/.kube/config', 'exec', '-i', 'ubuntu', '-c', 'ubuntu', '--', '/bin/sh', '-c', "/bin/sh -c 'echo ~ && sleep 0'"]
<127.0.0.1> EXEC ['/usr/bin/bash', '-c', 'EKS_TOKEN=test', '', '&&', '', '/usr/local/bin/kubectl', '-n', 'test', '--kubeconfig', '/.kube/config', 'exec', '-i', 'ubuntu', '-c', 'ubuntu', '--', '/bin/sh', '-c', '/bin/sh -c \'echo "`pwd`" && sleep 0\'']
<127.0.0.1> EXEC ['/usr/bin/bash', '-c', 'EKS_TOKEN=test', '', '&&', '', '/usr/local/bin/kubectl', '-n', 'test', '--kubeconfig', '/.kube/config', 'exec', '-i', 'ubuntu', '-c', 'ubuntu', '--', '/bin/sh', '-c', '/bin/sh -c \'( umask 77 && mkdir -p "` echo ~/.ansible/tmp `"&& mkdir "` echo ~/.ansible/tmp/ansible-tmp-1713438683.4774194-5211-164264421022768 `" && echo ansible-tmp-1713438683.4774194-5211-164264421022768="` echo ~/.ansible/tmp/ansible-tmp-1713438683.4774194-5211-164264421022768 `" ) && sleep 0\'']
fatal: [localhos

```

**Comment by avelytchko** [comment]
> ```yaml
> ansible_connection: kubectl
> ```

It just doesn't work as plugin requires successful connection BEFORE and pass any tasks and evn vars only AFTER connection is successful. But AWS env vars are required for connection so it's failed. We just received unreachable. Any task will fail immediately.

**Comment by yurnov** [comment]
Hi @gravesm,

it works in general cases, and TEST_ENV will be populated into target container __after connection__, but my ticket is for __local__ (not __target__) environmental variable that requried for esbalishing connection and need to be set on local (ansible controller) __before connection__ will be established in general case. That environmental variable required for authentication for connection isfelf.

**Comment by gravesm** [comment]
Sorry, maybe I'm not understanding your problem. The connection plugin is just spawning a `kubectl` child process and will inherit whatever environment your initial ansible process has. If you just need, for example, `AWS_ACCESS_KEY_ID` in the environment that the `kubectl` process is run in, just make sure it's set in whatever environment you are running `ansible-playbook`.

**Comment by yurnov** [comment]
Hi @gravesm, you are correct, this connection plugin just uses `kubectl` binary, and the local env variable is inherited, and playbook like that:
```
- hosts: cluster1:cluster2
  gather_facts: no
  any_errors_fatal: yes
  vars:
    ansible_connection: "kubernetes.core.kubectl"
    ansible_kubectl_namespace: "test"
    ansible_kubectl_config: "{{ kubeconfig }}"
    ansible_kubectl_pod: "shell"
    ansible_kubectl_container: "shell"
  environment:
    TEST_ENV1: value1
    TEST_ENV2: value2

  tasks:
  - name: test
    ansible.builtin.shell: env
    register: result
  - debug:
      var: result.stdout_lines
```
works fine when I have local environment variable required for `kubectl`, i.e. `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` that mentioned earlier, or any other that is required by kubectl in specific case. But it works perfectly when I have only one host (kubernetes cluster in case on `kubernetes.core.kubectl`, but if I have a few hosts with different credentials, it will not work.

Please have an example:
```
root@ubuntu-shell:/# cat inventory.ini
[k8s]
foo.example.com ansible_connection=kubernetes.core.kubectl  kubeconfig=/root/.kube/foo.example.com.config AWS_ACCESS_KEY_ID=foo.id AWS_SECRET_ACCESS_KEY=foo.key
bar.example.com  ansible_connection=kubernetes.core.kubectl  kubeconfig=/root/.kube/bar.example.com.config AWS_ACCESS_KEY_ID=bar.id AWS_SECRET_ACCESS_KEY=bar.key
root@ubuntu-shell:/# cat playbook.yaml
---
- hosts: cluster1:cluster2
  gather_facts: no
  any_errors_fatal: yes
  vars:
    ansible_connection: "kubernetes.core.kubectl"
    ansible_kubectl_namespace: "test"
    ansible_kubectl_config: "{{ kubeconfig }}"
    ansible_kubectl_pod: "shell"
    ansible_kubectl_container: "shell"
    ansible_kubectl_local_env:
       AWS_ACCESS_KEY_ID: "{{ aws_access_key_id }}"
       AWS_SECRET_ACCESS_KEY: "{{ aws_secret_access_key }}"
  environment:
    TEST_ENV1: value1
    TEST_ENV2: value2

  tasks:
  - name: test
    ansible.builtin.shell: env
    register: result
  - debug:
      var: result.stdout_lines
```

That will not work

**Comment by yurnov** [comment]
I tried to implement support of local env variable on my side, please find #699

**Comment by yurnov** [comment]
Hi @gravesm,

would you like to review PR #702 where the support of the local environmental variable requested in that issue was added? That was tested locally with some debug output, to do so `_connect` method modified to:
```python
    def _connect(self, port=None):
        """Connect to the container. Nothing to do"""
        super(Connection, self)._connect()
        if not self._connected:
            display.vvv(
              "ESTABLISH {0} CONNECTION".format(self.transport),
              host=self._play_context.remote_addr,
            )
            local_env = self._local_env()
            for key, value in local_env.items():
              display.vvv("ENV: {0}={1}".format(key, value), host=self._play_context.remote_addr)
            self._connected = True
```
(so, part
```
local_env = self._local_env()
            for key, value in local_env.items():
              display.vvv("ENV: {0}={1}".format(key, value), host=self._play_context.remote_addr)
```
was added)

Also, it was tested with AWS EKS and real environmental variables with AWS credentials that is my real case of why I created this issue.

 BR/Yuriy

**Comment by yurnov** [comment]
Hi @gravesm,

any chance to have PR reviewed soon?

BR/Yuriy

**Comment by yurnov** [comment]
PR #702 is merged to `main` and backported to `stable-3` as #705, so, it will be included in version 3.1.0 or any other minor further release. 

Thanks, @gravesm and @abikouo for their review and support

**Comment by yurnov** [comment]
Added in release [3.1.0](https://github.com/ansible-collections/kubernetes.core/releases/tag/3.1.0) by the PR #702

Changelog is [here](https://github.com/ansible-collections/kubernetes.core/blob/ef829b8a56787fa6c91dfe44a3775adc6cc1721d/changelogs/changelog.yaml#L844)

Closing ticket as completed
